lightwood/analysis/nn_conf/plinear.py

```diff
@@ -0,0 +1,167 @@
+    true and predicted values. Confidence score and bounds can be extracted based on repeated
+    sampling / bootstraping.
+
+    Given (X, yhat) -> learn param distributions to model residuals |y - yhat|
+    """
+    def __init__(self):
+        super().__init__()
+        self.model = None
+        self.trainer = None
+        self.optimizer = None
+        self.time_budget = 30
+        self.n_samples = 100  # 100 is the minimum for arbitrary quantiles at 0.01 resolution
+        self.device, _ = get_devices()
+        self.pct_train = 0.8
+        self.pct_test = 1 - self.pct_train
+        self.callback = lambda err, y, z: log.debug(f'PLinear error is {err}')
+        assert self.pct_train + self.pct_test == 1.0
+
+    def analyze(self, info: Dict[str, object], **kwargs) -> Dict[str, object]:
+        ns = SimpleNamespace(**kwargs)
+
+        X = ns.encoded_val_data.get_encoded_data(include_target=False)
+        y = torch.tensor(ns.encoded_val_data.get_column_original_data(ns.target).values).unsqueeze(1)
+        yh = torch.tensor(ns.normal_predictions['prediction'].values).unsqueeze(1)
+        r = (y - yh).float()
+        X = torch.hstack([X, yh]).float()  # predicted value is the most important prior
+
+        mask = torch.rand((len(y),)) < 0.8
+        train_X = X[mask, :]
+        test_X = X[~mask, :]
+        train_y = r[mask]
+        test_y = r[~mask]
+
+        n_targets = 1
+        n_features = X.shape[1]
+        self.model = PLinear(in_features=n_features, out_features=n_targets)
+        self.optimizer = torch.optim.Adam(params=self.model.parameters())
+        loss_fn = torch.nn.MSELoss()
+        scheduler = None
+
+        self.trainer = Gym(self.model, self.optimizer, scheduler, loss_fn, self.device)
+        train_data_loader = DataLoader(list(zip(train_X, train_y)),
+                                       batch_size=min(32, len(train_X)),
+                                       shuffle=True)
+        test_data_loader = DataLoader(list(zip(test_X, test_y)),
+                                      batch_size=min(32, len(test_X)),
+                                      shuffle=True)
+
+        self.trainer.fit(train_data_loader, test_data_loader,
+                         desired_error=0.01,
+                         max_time=self.time_budget,
+                         callback=self.callback)
+
+        return info
+
+    def explain(self,
+                row_insights: pd.DataFrame,
+                global_insights: Dict[str, object], **kwargs) -> Tuple[pd.DataFrame, Dict[str, object]]:
+
+        ns = SimpleNamespace(**kwargs)
+        conf_level = ns.pred_args.fixed_confidence
+        if conf_level is None:
+            conf_level = 0.95
+
+        X = ns.encoded_data
+        yh = torch.tensor(ns.predictions['prediction'].values).unsqueeze(1)
+        X = torch.hstack([X, yh]).float()
+
+        samples = []
+        for _ in range(self.n_samples):
```

Comment:
Just tried this out. There is a steep memory-time trade off here, because samples are not in the batch dimension: each data point in the batch should be sampled `n_samples` times. 

We can either increase the size of the batch `n_samples` times and manually pack/unpack the result (which could get out of hand for big datasets where the validation dataset is `10^5` or higher), or keep the iterator.

Only because we are not currently enforcing a maximum validation dataset size, I'm tempted to keep the current approach to avoid OOM issues. Thoughts?